Migrate a single node_authorizer.go klog.Infof call to klog.InfoS

ankeesler opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

We are starting to migrate the most popular log lines to use structured logging.

Ref:
* Structured logging KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging
* Structured logging work tracking: https://github.com/kubernetes/enhancements/issues/1602

**Special notes for your reviewer**:

I assumed that I should only move over the single `NODE DENY` log line as referenced in https://github.com/kubernetes/enhancements/issues/1602#issue-577135877; I figured if I was wrong about that assumption, I could hear about it after I opened a PR.

```release-note
NONE
```

ankeesler committed:
Migrate a single node_authorizer.go klog.Infof call to klog.InfoS

We are starting with the log lines that show up most often.

Signed-off-by: Andrew Keesler <akeesler@vmware.com>

serathius commented:
/lgtm
/assign @deads2k

serathius commented:
/ok-to-test

ankeesler commented:
/retest

ankeesler commented:
/retest

ankeesler commented:
/retest

serathius commented:
/retest

serathius commented:
ping @deads2k

serathius commented:
@liggitt can you take a look?

liggitt commented on plugin/pkg/auth/authorizer/node/node_authorizer.go:
```diff
@@ -206,7 +206,7 @@ func (r *NodeAuthorizer) authorize(nodeName string, startingType vertexType, att
 	ok, err := r.hasPathFrom(nodeName, startingType, attrs.GetNamespace(), attrs.GetName())
 	if err != nil {
-		klog.V(2).Infof("NODE DENY: %v", err)
+		klog.V(2).InfoS("NODE DENY", "err", err)
```
is "err" a standard name?

liggitt replied:
do you have example output from this change?

serathius replied:
Yes, more details: https://github.com/kubernetes/community/blob/5e71635bb60dd7684e1afd820a1a61b9dd975211/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#name-arguments

ankeesler replied:
> is "err" a standard name?

it seems to be. i used the "err" name because @serathius recommended that elsewhere: https://github.com/kubernetes/kubernetes/pull/91576#discussion_r432474929.

> do you have example output from this change?

ah, yeah, sorry, i forgot to put the output from this change in this pr.

before:
```
I0603 09:59:27.676862   54009 node_authorizer.go:209] NODE DENY: node "node2" cannot get secret ns1/node1-only, no relationship to this object was found in the node authorizer graph
```

after:
```
I0603 09:57:29.759232   53238 node_authorizer.go:209] "NODE DENY" err=&errors.errorString{s:"node \"node2\" cannot get secret ns1/node1-only, no relationship to this object was found in the node authorizer graph"}
```

serathius replied:
@ankeesler Can you use instructions provided herehttps://github.com/kubernetes/community/blob/5e71635bb60dd7684e1afd820a1a61b9dd975211/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#verify-log-output 
To generate comparison of log output?

liggitt replied:
so for existing log consumers, this results in much less readable output:

> I0603 09:59:27.676862   54009 node_authorizer.go:209] NODE DENY: node "node2" cannot get secret ns1/node1-only, no relationship to this object was found in the node authorizer graph

vs

> I0603 09:57:29.759232   53238 node_authorizer.go:209] "NODE DENY" err=&errors.errorString{s:"node \"node2\" cannot get secret ns1/node1-only, no relationship to this object was found in the node authorizer graph"}

serathius replied:
This looks like a bug. We intended to use `%q` for errors.
Filed: https://github.com/kubernetes/klog/issues/158

ankeesler replied:
Is there action that needs to be taken on this PR in the interim before the bug is fixed?

I could pass `err.Error()` instead of `err`?

serathius replied:
We should wait until bug is fixed. We shouldn't pass `err.Error()` as it drops information about type that could be used by `InfoS` implementation.

serathius replied:
@ankeesler I have created PR with fix to serialization. https://github.com/kubernetes/kubernetes/pull/91792
Please to generate output with this change

ankeesler replied:
Using @serathius's branch from the above PR, here is what the output diff is.

`Infof`:
```
I0605 14:45:43.330939   41909 node_authorizer.go:209] NODE DENY: node "node2" cannot get secret ns1/node1-only, no relationship to this object was found in the node authorizer graph
```

`InfoS`:
```
I0605 14:43:40.047434   41115 node_authorizer.go:209] "NODE DENY" err="node \"node2\" cannot get secret ns1/node1-only, no relationship to this object was found in the node authorizer graph"
```

liggitt replied:
the quote escaping is unfortunate... I didn't really expect non-json log output to change here, but if it's going to, we should avoid quoting things that will be escaped in ugly ways

serathius replied:
Quoting in InfoS here is needed to ensure logs can be parsed. I would be useful to move information stored in errors to logs (like node name). For this PR I think we should remove quotes used in error. 
@ankeesler can you change original error to not add quotes?

liggitt replied:
> Quoting in InfoS here is needed to ensure logs can be parsed

when output as JSON, sure... when output as text, I didn't expect so

ankeesler replied:
After 0715ecfbb96, the structured log line looks like this.
```
I0605 15:14:45.707400   53254 node_authorizer.go:209] "NODE DENY" err="node node2 cannot get secret ns1/node1-only, no relationship to this object was found in the node authorizer graph"
```

Note - there was a pattern[0][1][2][3][4] of quoting the node name in log lines from this file. The log lines could potentially have different variations of quoting/not quoting the node name now. Is this OK?

[0] https://github.com/kubernetes/kubernetes/blob/b8b4186a14045ab66b150b5a92276d02b8a73a3e/plugin/pkg/auth/authorizer/node/node_authorizer.go#L210
[1] https://github.com/kubernetes/kubernetes/blob/b8b4186a14045ab66b150b5a92276d02b8a73a3e/plugin/pkg/auth/authorizer/node/node_authorizer.go#L213
[2] https://github.com/kubernetes/kubernetes/blob/b8b4186a14045ab66b150b5a92276d02b8a73a3e/plugin/pkg/auth/authorizer/node/node_authorizer.go#L238
[3] https://github.com/kubernetes/kubernetes/blob/b8b4186a14045ab66b150b5a92276d02b8a73a3e/plugin/pkg/auth/authorizer/node/node_authorizer.go#L308
[4] https://github.com/kubernetes/kubernetes/blob/b8b4186a14045ab66b150b5a92276d02b8a73a3e/plugin/pkg/auth/authorizer/node/node_authorizer.go#L313

liggitt replied:
I guess replace those with `'%s'`?

ankeesler replied:
Gave that a go in 1b7352645bc.

liggitt replied:
I would still wrap in single-quotes so it is clear which part is the name

liggitt replied:
> here is where we are at with Infof logging.
>
> I0605 16:14:30.996413   76058 node_authorizer.go:281] NODE DENY: 'node0' authorizer.AttributesRecord{User:(*user.DefaultInfo)(0xc0003b16c0), Verb:"watch", Namespace:"", APIGroup:"storage.k8s.io", APIVersion:"", Resource:"csinodes", Subresource:"", Name:"", ResourceRequest:true, Path:""}

not sure I understand that... under what circumstances are we stringifying an interface pointer?

liggitt replied:
edit: nevermind, those were two different lines. switching the quote to single quote looks fine to me

serathius commented:
/retest

ankeesler commented:
/retest

ankeesler committed:
Remove quotes from error for readability

Signed-off-by: Andrew Keesler <akeesler@vmware.com>

ankeesler commented:
/retest

ankeesler committed:
node_authorizer.go: use %s for node names for log uniformity

Signed-off-by: Andrew Keesler <akeesler@vmware.com>

ankeesler committed:
node_authorizer.go: single-quote node name for readability++

This is good because:
1) the node name is clear in the log line
2) the node names shows up the same in {un-,}structured logs

Signed-off-by: Andrew Keesler <akeesler@vmware.com>

ankeesler commented:
/retest

serathius commented:
It looks strange that you not only change change quotes in fmt.Errorf, but also in klog.Infof. I think it would be better just to change them to klog.InfoS.
I leave the decision to @liggitt

liggitt commented:
> It looks strange that you not only change change quotes in fmt.Errorf, but also in klog.Infof. I think it would be better just to change them to klog.InfoS.

I think this is fairly representative of the types of changes I would expect us to make to keep structured and unstructured log calls consistent and as readable as possible.

serathius commented:
Is change ok for approval? @liggitt

liggitt commented:
/lgtm
/approve

serathius commented:
Don't see any other comments, should be ok to lgtm?

serathius commented:
/lgtm

serathius commented:
/sig instrumentation

serathius commented:
/priority important-soon
